src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java

```diff
@@ -495,4 +524,83 @@ public static Path inputManifestPath(Path runfilesDir) {
   public static Path outputManifestPath(Path runfilesDir) {
     return runfilesDir.getRelative(OUTPUT_MANIFEST_BASENAME);
   }
+
+  @Nullable
+  private static Artifact createRepoMappingManifestAction(
+      RuleContext ruleContext, Runfiles runfiles, Artifact owningExecutable) {
+    NestedSet<Package> transitivePackages =
+        ruleContext.getTransitivePackagesForRunfileRepoMappingManifest();
+    if (transitivePackages == null) {
+      // For environments where transitive packages are not tracked, we don't have external repos,
+      // so don't build the repo mapping manifest in such cases.
+      return null;
+    }
+
+    ImmutableList<Entry> entries = collectRepoMappings(transitivePackages, runfiles);
+    if (entries == null) {
+      // If Bzlmod is not enabled, nobody reads the repo mapping manifest anyway, so don't generate
+      // it.
+      return null;
+    }
+
+    PathFragment executablePath =
+        (owningExecutable != null)
+            ? owningExecutable.getOutputDirRelativePath(
+                ruleContext.getConfiguration().isSiblingRepositoryLayout())
+            : ruleContext.getPackageDirectory().getRelative(ruleContext.getLabel().getName());
+    Artifact repoMappingManifest =
+        ruleContext.getDerivedArtifact(
+            executablePath.replaceName(executablePath.getBaseName() + REPO_MAPPING_MANIFEST_EXT),
+            ruleContext.getBinDirectory());
+    ruleContext
+        .getAnalysisEnvironment()
+        .registerAction(
+            new RepoMappingManifestAction(
+                ruleContext.getActionOwner(),
+                repoMappingManifest,
+                entries,
+                ruleContext.getWorkspaceName()));
+    return repoMappingManifest;
+  }
+
+  /**
+   * Returns the list of entries (unsorted) that should appear in the repo mapping manifest, or null
+   * if Bzlmod is not enabled.
+   */
+  @Nullable
+  private static ImmutableList<Entry> collectRepoMappings(
+      NestedSet<Package> transitivePackages, Runfiles runfiles) {
+    ImmutableSet<RepositoryName> reposContributingRunfiles =
+        runfiles.getAllArtifacts().toList().stream()
+            .filter(a -> a.getOwner() != null)
+            .map(a -> a.getOwner().getRepository())
+            .collect(toImmutableSet());
+    Set<RepositoryName> seenRepos = new HashSet<>();
+    ImmutableList.Builder<Entry> entries = ImmutableList.builder();
+    // We use a rather hacky way to detect whether Bzlmod is enabled here: there exist repos that
+    // use strict deps iff Bzlmod is enabled. So if none of the repos we come upon use strict deps,
+    // either Bzlmod is disabled, or we're using a very weird subset of packages that *only* come
+    // from WORKSPACE-repos. In either case, it's unnecessary to output the repo mapping manifest.
+    boolean areWeSureBzlmodIsEnabled = false;
```

Comment:
It looks like we could simply call this `needsRepoMapping` and get rid of the comment, unless it helps Googlers monitoring action counts sleep soundly :-)

Reply:
Done.